### ISSUE TITLE
service-worker(safari): POST requests getting blocked with 504 by Service Worker

### ISSUE THREAD

**Issue by Splaktar** [angular/angular#37133]
# 🐞 bug report

### Affected Package
`@angular/service-worker`

### Is this a regression?
I don't know if this has even worked properly in Safari, but it's supposed to be production ready.

### Description
Service Workers shouldn't block POST requests AFAIK, but that's happening on Safari `Version 13.1 (15609.1.20.111.8)` on macOS Catalina `10.15.4 (19E287)`.

## 🔬 Minimal Reproduction
- Go to https://angular.io/ in Safari on macOS
- Open your Dev tools Console
- Browse around to some different pages like resources, guides, events, etc
- Observe the following error in the console
```
https://www.google-analytics.com/j/collect?v=1&_v=j82&a=557396380&t=pageview&_s=1&dl=https%3A%2F%2Fangular.io%2Fresources%3Fcategory%3Ddevelopment&dp=%2Fresources&ul=en&de=UTF-8&dt=Angular&sd=24-bit&sr=1920x1080&vp=1914x532&je=1&_u=QACAAEABAAAAAC~&jid=43561559&gjid=590323957&cid=498157216.1589533208&tid=UA-8594346-15&_gid=406270144.1589533208&_r=1&z=73615628
Failed to load resource: the server responded with a status of 504 (Gateway Timeout)
```

## 🔥 Exception or Error
```
Cannot use wildcard in Access-Control-Allow-Origin when credentials flag is true., undefined
```

<pre><code>
NGSW Debug Info:

Driver state: NORMAL ((nominal))
Latest manifest hash: a51465926eaa708ba14e763b3c654f2dc464f776
Last update check: 1m3s560u

=== Version a51465926eaa708ba14e763b3c654f2dc464f776 ===

Clients: 1692-8, 1692-33, 1692-59, 1692-63, 1692-70

=== Idle Task Queue ===
Last update tick: 28s363u
Last update run: 1m3s868u
Task queue:

Debug log:

[1m31s277u] TypeError(Cannot use wildcard in Access-Control-Allow-Origin when credentials flag is true., undefined) Driver.fetch(https://www.google-analytics.com/j/collect?v=1&_v=j82&a=1196111008&t=pageview&_s=1&dl=https%3A%2F%2Fangular.io%2Fguide%2Fservice-worker-communications&dp=%2Fguide%2Fservice-worker-communications&ul=en&de=UTF-8&dt=Angular&sd=24-bit&sr=1920x1080&vp=1914x532&je=1&_u=QACAAEABAAAAAC~&jid=323136214&gjid=1958337307&cid=498157216.1589533208&tid=UA-8594346-15&_gid=406270144.1589533208&_r=1&z=559832955)
[28s598u] TypeError(Cannot use wildcard in Access-Control-Allow-Origin when credentials flag is true., undefined) Driver.fetch(https://www.google-analytics.com/j/collect?v=1&_v=j82&a=1715609439&t=pageview&_s=5&dl=https%3A%2F%2Fangular.io%2Fguide%2Faccessibility&dp=%2Fresources&ul=en&de=UTF-8&dt=Angular%20-%20EVENTS&sd=24-bit&sr=1920x1080&vp=1914x532&je=1&_u=SACAAEABAAAAAC~&jid=721523635&gjid=563131875&cid=498157216.1589533208&tid=UA-8594346-15&_gid=406270144.1589533208&_r=1&z=882809832)
...
[3m52s] TypeError(Cannot use wildcard in Access-Control-Allow-Origin when credentials flag is true., undefined) Driver.fetch(https://www.google-analytics.com/j/collect?v=1&_v=j82&a=1121528947&t=pageview&_s=1&dl=https%3A%2F%2Fangular.io%2Fresources%3Fcategory%3Ddevelopment&dp=%2Fresources&ul=en&de=UTF-8&dt=Angular&sd=24-bit&sr=1920x1080&vp=1914x523&je=1&_u=QACAAEABAAAAAC~&jid=1658082190&gjid=2097738465&cid=498157216.1589533208&tid=UA-8594346-15&_gid=406270144.1589533208&_r=1&z=747985136)
[52s802u] TypeError(Cannot use wildcard in Access-Control-Allow-Origin when credentials flag is true., undefined) Driver.fetch(https://www.google-analytics.com/j/collect?v=1&_v=j82&a=1121528947&t=pageview&_s=3&dl=https%3A%2F%2Fangular.io%2Fresources%3Fcategory%3Ddevelopment&dp=%2Ffeatures&ul=en&de=UTF-8&dt=Angular%20-%20Angular%20Contributors&sd=24-bit&sr=1920x1080&vp=1914x523&je=1&_u=SACAAEABAAAAAC~&jid=2000655573&gjid=1284227509&cid=498157216.1589533208&tid=UA-8594346-15&_gid=406270144.1589533208&_r=1&z=1228242795)
</code></pre>

![Screen Shot 2020-05-15 at 05 28 50](https://user-images.githubusercontent.com/3506071/82035040-02c09a80-966d-11ea-9072-835431996a37.png)

![Screen Shot 2020-05-15 at 05 12 17](https://user-images.githubusercontent.com/3506071/82033899-6649c880-966b-11ea-9ffd-b434a272bd48.png)

## 🌍  Your Environment

**Angular Version:** `10.0.0-next.5`

**Anything else relevant?**
I'm also seeing this on my site when using LogRocket (another kind of analytics service), but the Headers it complains about are different:

```ts
Request header field Pragma is not allowed by Access-Control-Allow-Headers., undefined
```

```
NGSW Debug Info:

Driver state: EXISTING_CLIENTS_ONLY (Degraded due to failed initialization: Internal error
undefined)
Latest manifest hash: 5642c5e828814d3dae722eae1783dc4724682328
Last update check: 4s519u

=== Version 5642c5e828814d3dae722eae1783dc4724682328 ===

Clients: 779-37

=== Idle Task Queue ===
Last update tick: 3s197u
Last update run: 46s658u
Task queue:
 * check-updates-on-navigation

Debug log:
[5m51s257u] TypeError(Request header field Pragma is not allowed by Access-Control-Allow-Headers., undefined) Driver.fetch(https://r.lr-ingest.io/i?a=buymic%2Fdevintent-website&r=4-e8456afb-9941-460d-a17c-fddff8c6183e&t=e8d3692a-bb84-48ad-8f91-7dd31c74115c&ir=f&ht=f&s=0)
[5m48s165u] TypeError(Request header field Pragma is not allowed by Access-Control-Allow-Headers., undefined) Driver.fetch(https://r.lr-ingest.io/i?a=buymic%2Fdevintent-website&r=4-e8456afb-9941-460d-a17c-fddff8c6183e&t=e8d3692a-bb84-48ad-8f91-7dd31c74115c&ir=f&ht=f&s=0)
```
![Screen Shot 2020-05-15 at 05 19 27](https://user-images.githubusercontent.com/3506071/82034087-b58ff900-966b-11ea-9119-c9186bcd428c.png)
![Screen Shot 2020-05-15 at 05 21 06](https://user-images.githubusercontent.com/3506071/82034225-e8d28800-966b-11ea-858d-8a3c35ffa242.png)
![Screen Shot 2020-05-15 at 05 20 46](https://user-images.githubusercontent.com/3506071/82034230-e96b1e80-966b-11ea-84fe-6ef49684c35d.png)

**Comment by gkalpak** [comment]
This needs further investigation. Note that it might be related to the SW's stripping off request headers (see #24227 for more info).

**Comment by kosso** [comment]
I am also getting spurious 504 errors, only on Safari, but with some GET requests.

I never thought it might be the Service Worker...

**Comment by johannesjo** [comment]
I am encountering [the same issue](https://github.com/johannesjo/super-productivity/issues/645), which is a big problem because currently `ngsw-bypass=true` is the only way to exclude requests (see https://github.com/angular/angular/issues/21191).

**Comment by DibyodyutiMondal** [comment]
This is a big problem when using google tag manager.
All requests made from the page are going through the service worker.
Since the urls point to outside the service worker scope, all requests to Facebook Pixel, Google Analytics, etc are getting a CORS issue, causing the fetch to fail.

These requests are not being made by us - we load libraries that then make the requests, when events such as 'purchase' happens. As it is, we can't edit third party libraries that we requesting at runtime to include `ngsw-bypass` in their requests.
Even more so when that request is being delegated to Google Tag Manager.

I would recommend adding a simple way that allows us to set origins/urls via `ngsw-config.json` that are simply handed over to the browser and the service worker ignores it.

**Comment by kosso** [comment]
@DibyodyutiMondal  Have you tried adding an entry to the `dataGroup` array in `ngsw-config.json` with `maxSize` and `maxAge` of zero? 

```
{
      "name":"dont_use_sw",
      "urls": ["/some-matched-url"],
      "cacheConfig": {
        "maxSize": 0,
        "maxAge": "0u",
        "strategy": "freshness"
      }
}, 
...
```

You could also try a 'request interceptor' which could automatically add a `ngsw-bypass` query parameter to any ULRs you define, which should *always* work to bypass the Service Worker.

**Comment by DibyodyutiMondal** [comment]
I have tried it the datagroup method - it doesn't work for the scenario I described. Because the url in question is not part of my domain. It's part of the third-party domains...
For example
- `https://connect.facebook.net/fbevents.js` (javascript file)
- `https://connect.facebook.net/signals...` (simple api get request for metrics, not a file)
- `https://www.google-analytics/g/collect...` (simple api get request for metrics, not a file)

The only external url that is working properly is `gtag` (google tag manager) itself because that is being invoked directly in index.html. It loads a javascript library and the other urls are being requested dynamically by the gtag script. They obviously don't have the same origin, but because the service-worker intercepts them, they all start seeing cors errors.

Simplest way would be to just be able to specify to service worker to ignore such URLs.
In the current service worker, adding any of the urls to my `ngsw-config` automatically adds a starting `/`, because of a `adapter.normalizeUrl` call in the service worker.
 
I don't know if the interceptor method will work because these requests are being made by the libraries that are being loaded by google tag manager, not my angular code. Even if it did, it seems to me very heavy handed and I'd additionally also have to configure my analytics streams to ignore `ngsw-bypass` if they saw it - that is if they even allow such kind of configuration.

However, I'll try out the interceptor method and update the thread.

**Comment by DibyodyutiMondal** [comment]
> However, I'll try out the interceptor method and update the thread.

The interceptor method did not work. The interceptor is not receiving them.

**Comment by gkalpak** [comment]
@DibyodyutiMondal, this sounds like a different problem that what this issue was originally about. Please open a new issue.

**Comment by muuvmuuv** [comment]
I think this needs a more convenient way since not all SDK's allow modifying/adding query parameters, e.g. Sentry does no longer and Google Analytics for events as well.

Can we not just have a `bypassGroups` in "ngsw-config.json"? That would make it a lot easier to configure with regex.

**Comment by nandowalter-lm** [comment]
Hi guys, this seems the same as #53438 
I solved bypassing the sw for all requests of external resources
```
self.addEventListener('fetch', event => {
    if (!event?.request?.url.startsWith(self.location.origin)) {
        event.stopImmediatePropagation();
    }
});

importScripts('./ngsw-worker.js');
````
Anyone has found a better solution? Hope this can help you.

**Comment by edvinv** [comment]
I also have the same problem. Google Tag Manager is not working out box with Service workers on iPhone. GTM requests are blocked with 504 response. 
There is also no way to set ngsw-bypass parameter to scripts that are dynamically loaded, so the only solution is to replace  'ngsw-worker.js' with custom script.
In my case I use following script:
```js
const urlsToBypass = [
  'www.googletagmanager.com',
  '.google-analytics.com',
];

// Catch `fetch` events and prevent specific ones from propagating to the Angular SW.
self.addEventListener('fetch', evt => {
  try {
    const url = evt.request.url.toLowerCase();

    if (urlsToBypass.some(part => url.includes(part))) {
      evt.stopImmediatePropagation();
    }
  } catch (error) { }
});

// Import the Angular SW to do the heavy-lifting.
self.importScripts('./ngsw-worker.js');
```

I don't like this because I would prefer to use original ngsw-worker.js. Best solution will be to investigate why this scripts are blocked on iPhone and if there is way to solve it.  Other solution is that ngsw-worker.js' supports some kind of options including list of urls to bypass.

**Comment by vytautas-pranskunas-** [comment]
I have different error but also from SW
"405 method not allowed POST". After page refreash - everything starts to work. Error pops randomly.
Thats very annoying and painful for customers. 
I wonder when Angular tem will address that - this stays opened for 1.5 years!

I am getting this on Chrome!
